BraveShared/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,146 @@
+        completion?(favicon)
+        return
+      }
+
+      // Fetch bundled icons or custom icons first
+      bundledRenderer?.load(url: url) { [weak self] favicon in
+        guard let self = self, !self.isCancelled else {
+          completion?(nil)
+          return
+        }
+        
+        // A bundled or custom icon was found
+        if let favicon = favicon {
+          storeInCache(favicon, for: url, persistent: persistent)
+          completion?(favicon)
+          return
+        }
+        
+        // If there are no bundled or custom icons, fetch the icons from Brave-Core
+        faviconRenderer?.loadIcon(for: url, persistent: persistent) { [weak self] favicon in
+          guard let self = self, !self.isCancelled else {
+            completion?(nil)
+            return
+          }
+          
+          // A favicon was fetched
+          if let favicon = favicon {
+            storeInCache(favicon, for: url, persistent: persistent)
+          }
+          
+          completion?(favicon)
+        }
+      }
+    }
+    
+    cancellable.execute(on: .main)
+    return cancellable
+  }
+  
+  public class Cancellable {
+    private var workItem: DispatchWorkItem?
+    private var task: (Cancellable, BundledFaviconImageRenderer, FaviconRenderer) -> Void
+    private var bundledRenderer: BundledFaviconImageRenderer?
+    private var faviconRenderer: FaviconRenderer?
+    
+    public var isCancelled: Bool {
+      workItem?.isCancelled ?? true
+    }
+    
+    public func cancel() {
+      workItem?.cancel()
+      workItem = nil
+    }
+    
+    fileprivate init(task: @escaping (Cancellable?, BundledFaviconImageRenderer?, FaviconRenderer?) -> Void) {
+      self.task = task
+      self.bundledRenderer = BundledFaviconImageRenderer()
+      self.faviconRenderer = FaviconRenderer()
+      
+      self.workItem = DispatchWorkItem { [weak self] in
+        task(self, self?.bundledRenderer, self?.faviconRenderer)
+      }
+    }
+    
+    fileprivate func execute(on queue: DispatchQueue) {
+      if let workItem = workItem {
+        queue.async(execute: workItem)
+      }
+    }
+  }
+  
+  private static func cacheURL(for url: URL) -> URL {
+    // Some websites still only have a favicon for the FULL url including the fragmented parts
+    // But they won't have a favicon for their domain
+    // In this case, we want to store the favicon for the entire domain regardless of query parameters or fragmented parts
+    // Example: `https://app.uniswap.org/` has no favicon, but `https://app.uniswap.org/#/swap?chain=mainnet` does.
+    return URLOrigin(url: url).url ?? url
+  }
+  
+  private static func storeInCache(_ favicon: Favicon, for url: URL, persistent: Bool) {
+    // Do not cache non-persistent icons
+    // Do not cache monogram icons
+    if !persistent && !favicon.isMonogramImage {
+      guard let data = try? JSONEncoder().encode(favicon) else {
+        return
+      }
```

Comment:
The initial expression seems to be wrong? The comment says "Do not cache non-persistant icons" but then checking if persistent is false to cache. I assume you want "if persistent and not a monogram"

These expressions can also be merged

`guard persistant, !favicon.isMonogramImage, let data = try? JSONEncoder().encode(favicon) else { return }`